### PR TITLE
Firewalld and TcpWrappers Controls

### PR DESCRIPTION
**Pull request by jburns12**
Added and updated controls that deal with firewalld and tcpwrappers that are dependent on the firewalld and etc_hosts_allow/deny resources.

These resources are available in InSpec version 1.40+.

### CODE REVIEW

File: controls/V-72049.rb [hunk 1]
```diff
@@ -69,7 +69,6 @@
 user must log on to their account first and then switch the user to the application
 account with the correct option to gain the account’s environment variables."
 
-  # @todo - test for values more restrictive than 077
   file_lines = command('grep -i -s umask /home/*/.*').stdout.split("\n")
```
Comment (trevor-vaughan) [inline controls/V-72049.rb]: This is going to explode the heck out of NFS and also miss home directories.

You should query the current user list using `getent passwd` and go from there probably.

To test this properly you should *probably* drop to a shell using `runuser` and actually run the `umask` command. There are a LOT of ways for this to go wrong and, in our case, we set everything globally in `/etc/profile.d` so this check wouldn't pick anything up.

File: controls/V-72047.rb [hunk 2]
```diff
@@ -61,21 +67,11 @@
 
 # chgrp root <directory>"
 
-  # @todo - add option for app group associated with dir?
   ww_dirs = command('find / -xdev -perm -002 -type d -exec ls -ld {} \;').stdout.split("\n")
```
Comment (trevor-vaughan) [inline controls/V-72047.rb]: This seems a little aggressive, I'll have to dig around for my code that is more comprehensive. It's in one of the checks already in place.

File: controls/V-72219.rb [hunk 3]
```diff
@@ -81,4 +106,18 @@
 Assurance List (CAL), this is a finding."
   tag "fix": "Update the host's firewall settings and/or running services to comply
 with the PPSM CLSA for the site or program and the PPSM CAL."
+
+  if service('firewalld').running?
+    describe firewalld do
+      FIREWALLD_SERVICES_DENY.each do |serv|
+        it { should_not have_service_enabled_in_zone(serv) }
```
Comment (trevor-vaughan) [inline controls/V-72219.rb]: These zones are completely arbitrary and I have no idea how to do this properly.

`iptables` also needs to be checked since a lot of people still manage that directly.

File: controls/V-72271.rb [hunk 4]
```diff
@@ -67,7 +67,6 @@
 # firewall-cmd --direct --add-rule ipv4 filter IN_public_allow 0 -p tcp -m limit
 --limit 25/minute --limit-burst 100  -j ACCEPT"
 
-  # @todo - firewall resource?
   describe command('firewall-cmd --direct --get-rule ipv4 filter IN_public_allow') do
```
Comment (trevor-vaughan) [inline controls/V-72271.rb]: Also need to check `iptables`

File: controls/V-72315.rb [hunk 5]
```diff
@@ -20,6 +20,54 @@
...
+TCPWRAPPERS_ALLOW = attribute(
+  'tcpwrappers_allow',
+  default: [
+    # Example
+    # { 'daemon' => 'ALL', 'client_list' => ['ALL'], 'options' => ['allow'] }
+  ],
+  description: "Allow rules from etc/hosts.allow."
+)
+
+TCPWRAPPERS_DENY = attribute(
```
Comment (trevor-vaughan) [inline controls/V-72315.rb]: These are probably going to fail with 'global already defined' warnings when using inherited profiles. They should not be constants and need to be contained within the `control`

File: controls/V-72315.rb [hunk 6]
```diff
@@ -64,30 +106,64 @@
...
+      end
+    end
+
+  else
+    describe package('tcp_wrappers') do
+      it { should be_installed }
+    end
+    TCPWRAPPERS_ALLOW.each do |rule|
+      describe etc_hosts_allow.where { daemon == rule['daemon'] } do
+        its('client_list') { should include rule['client_list'] }
```
Comment (trevor-vaughan) [inline controls/V-72315.rb]: Haven't checked but did you validate that the rules handle `DDQ` notation properly? `tcpwrappers` cannot handle `CIDR`.

File: controls/V-72219.rb (hunk 3 above)
Comment (jburns12) [reply controls/V-72219.rb]: This is testing services in the default zone by default...most of this firewall stuff seems like it will have to have the attributes configured manually depending on the needs/environment of the particular system, so I left all the attributes empty for now.

File: controls/V-72315.rb (hunk 5 above)
Comment (jburns12) [reply controls/V-72315.rb]: I'll change this. There are a few of these hanging out throughout the profile.

**Comment by trevor-vaughan** [comment]
@jburns12 Just checking in on this to see if you need any help or anything.

**Comment by aaronlippold** [comment]
Hi Trevor,

In the middle of getting my coverage back on the tasks for the next FY so I
am having Danny take over the task from Jen given she has to cover work for
DHS on the Attack framework. She and Danny are at the same site so they
should be able to collaborate on this.

Danny is also working on the scap_results_xml formatter as well so I think
this is a good pairing.

Collaboration Support Between the Teams:

Help with updating/writing the iptables resource to align with the
firewalld resource would be useful, help with rounding out some of these
corner cases I would like to share between the teams as I think your
ability - SIMP - to test some of these may be better. Suggestions on
'default' values for the attributes for the core-profile - which may be a
side effect of your inherited overlay profile with respect to the package
defaults for the services - would also be useful.

Also, if we could setup a weekly sync meeting until we hit 1.0 that might
help keep things moving if that makes sense.


--------
Aaron Lippold

lippold@gmail.com

260-255-4779

twitter/aim/yahoo,etc.
'aaronlippold'

File: controls/V-72315.rb (hunk 5 above)
Comment (aaronlippold) [reply controls/V-72315.rb]: @trevor-vaughan this is a known issue that I really need to get inspec to fix in terms of how they load and manage attributes in profiles. Perhaps you could reference the need in an issue with this example it would help to have an 'outside source'.

**Comment by jburns12** [comment]
Hey @trevor-vaughan and @aaronlippold ... I was able to at least address the addition of iptables for this PR with a few other fixes based on the review. The solutions for firewall controls are definitely not ideal, but I think it may take some time to map out what may be the best course of action for these in the long run.

File: controls/V-72047.rb [hunk 7]
```diff
@@ -20,6 +20,12 @@
 -----------------
 =end
 
+ww_dir_groups = attribute(
+  'ww_dir_groups',
+  default: ['root', 'sys', 'bin'],
+  description: "Groups world-writeable directories are allowed to be owned by"
```
Comment (trevor-vaughan) [inline controls/V-72047.rb]: This doesn't match the policy prose.

Application groups are < 500 on EL6 and < 1000 on EL7 systems by default.

Also, this should not be global.

File: controls/V-72049.rb [hunk 8]
```diff
@@ -69,12 +69,17 @@
...
-      its('stdout.strip') { should match /^umask\s+.*077/}
+
+  @user_params = passwd.params
+  @user_params.each do |user_param|
+    if user_param['uid'].to_i >= 1000
+      @file_lines = command("grep -i -s bash #{user_param['home']}/.*").stdout.split("\n")
+      @file_lines.each do |curr_line|
+        @file_name = curr_line.split(':').first
+        describe file(@file_name) do
+          its('content') { should_not match /^umask\s+[0-6]{1,3}$/ }
```
Comment (trevor-vaughan) [inline controls/V-72049.rb]: This needs to be restricted to only local accounts (in `/etc/passwd` and the `sssd` `LOCAL` domain if present) per the prose above. To do otherwise may cause issues on network mounted home directories.

File: controls/V-72219.rb [hunk 9]
```diff
@@ -20,6 +20,40 @@
...
+  'firewalld_ports_deny',
+  default: [
+    # Examples
+    # '12345/tcp',
+    # '23456/tcp'
+  ],
+  description: "Ports that firewalld should be configured to deny."
+)
+
+iptable_rules = attribute(
```
Comment (trevor-vaughan) [inline controls/V-72219.rb]: These make sense but probably should not be outside of the `control` scope

File: controls/V-72273.rb [hunk 10]
```diff
@@ -20,6 +20,12 @@
 -----------------
 =end
 
+FIREWALL_APPLICATION = attribute(
```
Comment (trevor-vaughan) [inline controls/V-72273.rb]: This should not be global and will cause issues with profiles.

File: controls/V-72315.rb [hunk 11]
```diff
@@ -20,6 +20,68 @@
 -----------------
 =end
 
+# These attributes must be filled in to reflect expectations of particular system
+FIREWALLD_SERVICES = attribute(
```
Comment (trevor-vaughan) [inline controls/V-72315.rb]: We should not have global items.

File: controls/V-72315.rb [hunk 12]
```diff
@@ -64,30 +120,69 @@
...
+        it { should have_rule(rule) }
+      end
+    end
+  else
+    describe package('tcp_wrappers') do
+      it { should be_installed }
+    end
+    tcpwrappers_allow.each do |rule|
+      describe etc_hosts_allow.where { daemon == rule['daemon'] } do
+        its('client_list') { should include rule['client_list'] }
```
Comment (trevor-vaughan) [inline controls/V-72315.rb]: This should probably be something like `be` instead of `include`. In theory, this should be a maxiumum set.

We're going to need a way to feed site specific data into these checks without modifying the files themselves because that's just not viable over time.

**Comment by trevor-vaughan** [comment]
@aaronlippold Is this going to be picked up by any one?

File: controls/V-72047.rb (hunk 2 above)
Comment (djhaynes) [reply controls/V-72047.rb]: @trevor-vaughan did you every find your code that should be used here?

File: controls/V-72271.rb (hunk 4 above)
Comment (djhaynes) [reply controls/V-72271.rb]: This has already been addressed in #33.

File: controls/V-72315.rb (hunk 12 above)
Comment (djhaynes) [reply controls/V-72315.rb]: @trevor-vaughan I changed these from "include" to "be". As far as site-specific data, can't you do that through the inputs that are defined in the control?

File: controls/V-72047.rb (hunk 2 above)
Comment (djhaynes) [reply controls/V-72047.rb]: I think this is actually all set. The current version of this control checks for world-writable directories that are not owned by root, sys, bin, or application groups.

**Comment by djhaynes** [comment]
Closed this PR since the changes have been re-submitted as separate PRs.
